packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -48,19 +48,43 @@ void joinCaches({
     final String newPath = fileSystem.path.join(targetDirectory.path, entity.basename);
     if (entity is File) {
       final File file = fileSystem.file(entity.path);
-      file.renameSync(newPath);
+      file.copySync(newPath);
+      file.deleteSync(recursive: true);
     }
     else if (entity is Directory) {
```

Comment:
style nit: this `else if` should be on the previous line, after the close curly brace